File: myExpenses/src/main/java/org/totschnig/myexpenses/util/ZipUtils.java
package org.totschnig.myexpenses.util;

import android.database.Cursor;
import android.net.Uri;
import android.os.Environment;
import android.text.TextUtils;

import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.provider.BackupUtilsKt;
import org.totschnig.myexpenses.provider.DatabaseConstants;
import org.totschnig.myexpenses.provider.TransactionProvider;
import org.totschnig.myexpenses.util.crypt.EncryptionHelper;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.security.GeneralSecurityException;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

import androidx.annotation.Nullable;
import androidx.documentfile.provider.DocumentFile;
import timber.log.Timber;


public class ZipUtils {


  public static final String PICTURES = Environment.DIRECTORY_PICTURES;

  private ZipUtils() {
  }

  public static void zipBackup(File cacheDir, DocumentFile destZipFile, String password)
      throws IOException, GeneralSecurityException {
    final OutputStream out = MyApplication.getInstance().getContentResolver().openOutputStream(destZipFile.getUri());
    ZipOutputStream zip = new ZipOutputStream(TextUtils.isEmpty(password) ? out : EncryptionHelper.encrypt(out, password));
    /*
     * add the folder to the zip
     */
    addFileToZip("", BackupUtilsKt.getBackupDbFile(cacheDir), zip);
    addFileToZip("", BackupUtilsKt.getBackupPrefFile(cacheDir), zip);
    Cursor c = MyApplication.getInstance().getContentResolver()
        .query(TransactionProvider.TRANSACTIONS_URI.buildUpon().appendQueryParameter(
            TransactionProvider.QUERY_PARAMETER_DISTINCT, "1").build(),
            new String[]{DatabaseConstants.KEY_PICTURE_URI},
            DatabaseConstants.KEY_PICTURE_URI + " IS NOT NULL",
            null, null);
    if (c != null) {
      if (c.moveToFirst()) {
        try {
          do {
            Uri imageFileUri = Uri.parse(c.getString(0));
            if (imageFileUri.getScheme().equals("file")) {
              File imageFile = new File(imageFileUri.getPath());
              if (imageFile.exists()) {
                addFileToZip(PICTURES, imageFile, zip);
              }
            } else {
              InputStream in;
              try {
                in = MyApplication.getInstance().getContentResolver().openInputStream(imageFileUri);
                addInputStreamToZip(PICTURES + "/" + imageFileUri.getLastPathSegment(),
                    in,
                    zip);
                in.close();
              } catch (FileNotFoundException e) {
                //File has been removed
                //Timber.e(e);
              }
            }
          } while (c.moveToNext());
        } catch (IOException e) {
          c.close();
          throw e;
        }
      }
      c.close();
    }
    /*
     * close the zip objects
     */
    zip.flush();
    zip.close();
  }

  /*
   * zip the folders
   */
  public static void zipFolder(File srcFolder, File destZipFile) throws Exception {
    ZipOutputStream zip;
    FileOutputStream fileWriter;
    /*
     * create the output stream to zip file result
     */
    fileWriter = new FileOutputStream(destZipFile);
    zip = new ZipOutputStream(fileWriter);
    /*
     * add the folder to the zip
     */
    addFolderToZip("", srcFolder, zip, true);
    /*
     * close the zip objects
     */
    zip.flush();
    zip.close();
  }

  /*
   * recursively add files to the zip files
   */
  private static void addFileToZip(String path, File srcFile,
                                   ZipOutputStream zip) throws IOException {

    FileInputStream in = new FileInputStream(srcFile);
    String filePath = path + (path.equals("") ? "" : "/") + srcFile.getName();
    addInputStreamToZip(filePath, in, zip);
    in.close();
  }

  private static void addInputStreamToZip(String path, InputStream in,
                                          ZipOutputStream zip) throws IOException {

    /*
     * write the file to the output
     */
    byte[] buf = new byte[1024];
    int len;
    zip.putNextEntry(new ZipEntry(path));
    while ((len = in.read(buf)) > 0) {
      /*
       * Write the Result
       */
      zip.write(buf, 0, len);
    }
    in.close();
  }

  /*
   * add folder to the zip file
   */
  private static void addFolderToZip(String path, File srcFolder, ZipOutputStream zip, boolean strip) throws IOException {

    /*
     * check the empty folder
     */
    if (srcFolder.list().length == 0) {
      zip.putNextEntry(new ZipEntry(path + "/" + srcFolder.getName() + "/"));
    } else {
      /*
       * list the files in the folder
       */
      String dirPath = strip ? "" : (path + (path.equals("") ? "" : "/") + srcFolder.getName());
      for (File file : srcFolder.listFiles()) {
        if (file.isDirectory()) {
          addFolderToZip(dirPath, file, zip, false);
        } else {
          addFileToZip(dirPath, file, zip);
        }
      }
    }
  }

  public static void unzip(InputStream fileIn, File dirOut, @Nullable String password)
      throws IOException, GeneralSecurityException {
    ZipInputStream zin = new ZipInputStream(TextUtils.isEmpty(password) ? fileIn : EncryptionHelper.decrypt(fileIn, password));
    ZipEntry ze;
    while ((ze = zin.getNextEntry()) != null) {
      Timber.v("Unzipping %s", ze.getName());
      File newFile = new File(dirOut, ze.getName());
      String canonicalPath = newFile.getCanonicalPath();
      if (!canonicalPath.startsWith(dirOut.getCanonicalPath())) {
        throw new SecurityException("Path Traversal Vulnerability");
      }
      newFile.getParentFile().mkdirs();
      if (ze.isDirectory()) {
        newFile.mkdir();
      } else {
        FileOutputStream fout = new FileOutputStream(newFile);
        long startTime = System.currentTimeMillis();

        byte[] buffer = new byte[1024];
        int count;
        while ((count = zin.read(buffer)) != -1) {
          fout.write(buffer, 0, count);
        }
        long endTime = System.currentTimeMillis();

        Timber.d("That took %d milliseconds", (endTime - startTime));
        zin.closeEntry();
        fout.close();
      }
    }
    zin.close();
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/util/ZipUtils.kt
package org.totschnig.myexpenses.util

import android.content.Context
import android.text.TextUtils
import androidx.core.net.toUri
import androidx.documentfile.provider.DocumentFile
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_ROWID
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_URI
import org.totschnig.myexpenses.provider.TransactionProvider
import org.totschnig.myexpenses.provider.asSequence
import org.totschnig.myexpenses.provider.fileName
import org.totschnig.myexpenses.provider.getBackupDataStoreFile
import org.totschnig.myexpenses.provider.getBackupDbFile
import org.totschnig.myexpenses.provider.getBackupPrefFile
import org.totschnig.myexpenses.util.crypt.EncryptionHelper
import timber.log.Timber
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.security.GeneralSecurityException
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

object ZipUtils {
    const val PICTURES = "Pictures"
    @Throws(IOException::class, GeneralSecurityException::class)
    fun zipBackup(
        context: Context,
        cacheDir: File?,
        destZipFile: DocumentFile,
        password: String?,
        lenientMode: Boolean,
    ) {
        val resolver = context.contentResolver
        val out = resolver.openOutputStream(destZipFile.uri)
        val zip = ZipOutputStream(
            if (TextUtils.isEmpty(password)) out else EncryptionHelper.encrypt(
                out!!, password
            )
        )
        addFileToZip("", getBackupDbFile(cacheDir), zip)
        addFileToZip("", getBackupPrefFile(cacheDir), zip)
        val backupDataStoreFile = getBackupDataStoreFile(cacheDir)
        if (backupDataStoreFile.exists()) {
            addFileToZip("", backupDataStoreFile, zip)
        }
        try {
            resolver
                .query(
                    TransactionProvider.ATTACHMENTS_URI,
                    arrayOf(KEY_ROWID, KEY_URI),
                    null,
                    null,
                    null
                )
        } catch (e: Exception) {
            if (lenientMode) null else throw e
        }?.use {
                it.asSequence.forEach { cursor ->
                    val rowId = cursor.getLong(0)
                    val uri = cursor.getString(1).toUri()
                    val fileName = "${rowId}_${uri.fileName(context)}"
                    try {
                        resolver.openInputStream(uri)?.use { inputStream ->
                            addInputStreamToZip(
                                "$PICTURES/$fileName",
                                inputStream,
                                zip
                            )
                        }
                    } catch (_: FileNotFoundException) {
                        //File has been removed
                        //Timber.e(e);
                    }
                }
            }
        zip.flush()
        zip.close()
    }

    /*
   * recursively add files to the zip files
   */
    @Throws(IOException::class)
    private fun addFileToZip(
        path: String, srcFile: File,
        zip: ZipOutputStream,
    ) {
        FileInputStream(srcFile).use {
            val filePath = path + (if (path == "") "" else "/") + srcFile.name
            addInputStreamToZip(filePath, it, zip)
        }
    }

    @Throws(IOException::class)
    private fun addInputStreamToZip(
        path: String, inputStream: InputStream,
        zip: ZipOutputStream,
    ) {
        val buf = ByteArray(1024)
        var len: Int
        zip.putNextEntry(ZipEntry(path))
        while (inputStream.read(buf).also { len = it } > 0) {
            zip.write(buf, 0, len)
        }
    }

    @Throws(IOException::class, GeneralSecurityException::class)
    fun unzip(fileIn: InputStream?, dirOut: File, password: String?) {
        val zin = ZipInputStream(
            if (TextUtils.isEmpty(password)) fileIn else EncryptionHelper.decrypt(
                fileIn,
                password
            )
        )
        while (true) {
            val ze = zin.nextEntry ?: break
            Timber.v("Unzipping %s", ze.name)
            val newFile = File(dirOut, ze.name)
            val canonicalPath = newFile.canonicalPath
            if (!canonicalPath.startsWith(dirOut.canonicalPath)) {
                throw SecurityException("Path Traversal Vulnerability")
            }
            newFile.parentFile.mkdirs()
            if (ze.isDirectory) {
                newFile.mkdir()
            } else {
                val fout = FileOutputStream(newFile)
                val startTime = System.currentTimeMillis()
                val buffer = ByteArray(1024)
                var count: Int
                while (zin.read(buffer).also { count = it } != -1) {
                    fout.write(buffer, 0, count)
                }
                val endTime = System.currentTimeMillis()
                Timber.d("That took %d milliseconds", endTime - startTime)
                zin.closeEntry()
                fout.close()
            }
        }
        zin.close()
    }
}